include/swift/SIL/SILInstruction.h

```diff
@@ -87,6 +88,121 @@ SILInstructionKind getSILInstructionKind(StringRef InstName);
 /// Map SILInstructionKind to a corresponding SILInstruction name.
 StringRef getSILInstructionName(SILInstructionKind Kind);
 
+/// A formal SIL reference to a list of values, suitable for use as the result
+/// of a SILInstruction.
+///
+/// *NOTE* Most multiple value instructions will not have many results, so if we
+/// want we can cache up to 3 bytes in the lower bits of the value.
+///
+/// *NOTE* Most of this defined out of line further down in the file to work
+/// around forward declaration issues.
+class SILInstructionResultArray {
+  const SILInstruction *ParentInst;
+
+public:
+  SILInstructionResultArray(const NonValueInstruction *NVI);
+  SILInstructionResultArray(const SingleValueInstruction *SVI);
+
+  SILValue operator[](size_t Index) const;
+
+  bool empty() const { return isa<NonValueInstruction>(ParentInst); }
+
+  size_t size() const;
+
+  class iterator {
+    friend class SILInstructionResultArray;
+
+    const SILInstructionResultArray &BaseArray;
+    Optional<unsigned> Index;
+
+    iterator(const SILInstructionResultArray &BaseArray,
+             Optional<unsigned> Index);
+    Optional<unsigned> getEndOffset() const;
+
+    bool valid() const;
+
+  public:
+    using value_type = SILValue;
+    using reference = SILValue;
+    using pointer = void;
+    using difference_type = ptrdiff_t;
+    using iterator_tag_iterator_category = std::random_access_iterator_tag;
+
+    SILValue operator*() const;
+
+    iterator &operator++() {
+      Index = *Index + 1;
+      assert(valid());
+      return *this;
+    }
+
+    iterator operator++(int) { return iterator(BaseArray, *Index + 1); }
+
+    bool operator==(iterator rhs) const {
+      return BaseArray.ParentInst == rhs.BaseArray.ParentInst &&
```

Comment:
You can assert this thing about BaseArray; iterators are not generally allowed to be compared to iterators from different collections.